### PR TITLE
Propose nft transaction

### LINKED ISSUE #62: [CONTRACT] Implement propose_nft_transaction Function
## Description 📝
We need to implement a `propose_nft_transaction` function for our protocol. This function will allow account members with proposer permission to create NFT transfer proposals that require approval from multiple signers before execution. 

The function will handle the creation and storage of NFT transfer transaction proposals, including essential details such as the NFT contract address, token ID, and recipient address. This functionality is crucial for multisignature accounts that need to manage and transfer NFTs collectively.

The implementation should leverage the existing `NFTTransaction` data structure and integrate with our multisignature approval workflow.

## Proposed Steps 📝
1. Create or update the `actions/nft_transaction.cairo` component to implement the `propose_nft_transaction` function with the following parameters:
   - `nft_contract`: The contract address of the NFT
   - `token_id`: The ID of the specific NFT token
   - `recipient`: The address to transfer the NFT to

2. Add validation logic to ensure:
   - The caller has proposer permission
   - The NFT contract address is valid and non-zero
   - The token ID is valid
   - The recipient address is valid, non-zero, and not the account itself
   - The account actually owns the specified NFT token

3. Integrate with the IERC721 interface to verify NFT ownership and approve future transfers

4. Generate a unique transaction ID for the proposal

5. Store the NFT transaction proposal in the contract's state

6. Emit an appropriate event when an NFT transaction is proposed

7. Implement functions to retrieve NFT transaction data:
   - `get_nft_transaction`: Get a specific NFT transaction by ID
   - `get_all_nft_transactions`: Get all NFT transactions

8. Update any related components or interfaces to support NFT transactions

9. Create comprehensive tests for the new functionality

## Acceptance Criteria ✅
- [ ] The `propose_nft_transaction` function is implemented in the appropriate component
- [ ] The function properly stores all required NFT transaction details (NFT contract, token ID, recipient)
- [ ] Appropriate validation checks are implemented to prevent invalid transactions:
  - [ ] Verifies proposer permission
  - [ ] Validates NFT contract address
  - [ ] Validates token ID
  - [ ] Validates recipient address
  - [ ] Confirms NFT ownership
- [ ] A unique transaction ID is generated for each proposal
- [ ] The function emits an event containing relevant details when an NFT transaction is proposed
- [ ] Data retrieval functions work correctly
- [ ] The function properly integrates with the existing multisignature workflow
- [ ] Comprehensive tests verify:
  - [ ] Valid NFT transactions can be proposed
  - [ ] Invalid NFT transactions are rejected with appropriate errors
  - [ ] Transactions without proper permissions are rejected
  - [ ] Events are properly emitted
- [ ] The implementation follows Cairo best practices and is gas-efficient
- [ ] Documentation is updated to explain the NFT transaction proposal process

## Additional Context 📌
This functionality is part of expanding our multisignature account system to handle various asset types beyond simple token transfers. After an NFT transaction is proposed using this function, it will go through the standard multisig approval and execution process.
Follow the Implementation of #58

### PR DESCRIPTION
**Pull request by FrankiePower**
**Pull Request: Implement NFT Transaction Proposal Functionality**

**Description 📝**  
This PR implements the `propose_nft_transaction` function in the `NFTTransaction` component, enabling multisignature account members with proposer permissions to create NFT transfer proposals. The implementation includes validation, storage, event emission, and retrieval functions (`get_nft_transaction` and `nft_transaction_list`), integrating with the existing multisignature workflow. Comprehensive tests verify the functionality, and documentation has been updated to explain the NFT transaction proposal process.

The work addresses the need to expand the multisignature account system to handle NFT transfers, ensuring secure and validated proposals that require multiple signers’ approval before execution.

**Related Issues 🔗**  
- Implements #58 (NFT Transaction Proposal Functionality)  
- Closes #62 

**Changes Made 🚀**  
- **✨ Feature Implementation**  
  - Added `propose_nft_transaction` to the `NFTTransaction` component with parameters `nft_contract`, `token_id`, and `recipient`.  
  - Implemented validation checks:  
    - Verifies caller has proposer permission using `PermissionControl::is_proposer`.  
    - Ensures `nft_contract` and `recipient` are non-zero addresses.  
    - Validates `token_id` is non-zero.  
    - Checks that `recipient` is not the account itself.  
    - Confirms account ownership of the NFT via `IERC721::owner_of`.  
  - Integrated with `AccountData` to generate a unique transaction ID and store the transaction with `TransactionType::NFT_SEND`.  
  - Stored NFT transaction data (`NFTTransactionData`) in a `Map<u256, NFTTransactionData>` and tracked IDs in a `Vec<u256>`.  
  - Emitted `NFTTransactionCreated` event with `id`, `nft_contract`, `token_id`, and `recipient`.  
  - Implemented `get_nft_transaction` to retrieve a specific NFT transaction by ID, validating transaction type.  
  - Implemented `nft_transaction_list` to return all NFT transactions.  
  - Updated `MockContract` to include the `NFTTransaction` component and expose public methods (`propose_nft_transaction_pub`, `get_nft_transaction_pub`, `nft_transaction_list_pub`).  

- **📚 Documentation Update**  
  - Updated documentation in `NFTTransaction.cairo` to explain the NFT transaction proposal process, validation logic, and integration with the multisignature workflow.  
  - Added comments in the test suite to clarify test scenarios and expected behavior.  

- **❓ Others**  
  - Added comprehensive tests to cover edge cases (e.g., non-existent token ID, invalid transaction type, non-existent transaction ID) and event emission.  
  - Ensured gas-efficient storage using `Map` and `Vec` for NFT transaction data and IDs.  

**Screenshots/Screen-record (if applicable) 🖼**  
Screenshot from 2025-05-03 22-43-59

**Checklist ✅**  
- 🛠 I have tested these changes using `scarb test` to ensure all tests in `NFTTransactionTests.cairo` pass.  
- 📖 I have updated the documentation in `NFTTransaction.cairo` to explain the NFT transaction proposal process.  
- 🎨 This PR follows the project’s Cairo coding style, using clear naming and modular component design.  
- 🧪 I have added necessary tests to cover success cases, failure cases, edge cases, and event emission.

### CODE REVIEW

File: src/actions/nft_transaction.cairo [hunk 1]
```diff
@@ -1,10 +1,389 @@
...
+            let erc721_dispatcher = IERC721Dispatcher { contract_address: nft_contract };
+            assert(erc721_dispatcher.owner_of(token_id) == account_address, Errors::ERR_NOT_OWNER);
+            // Check proposer permission
+            let permission_control_comp = get_dep_component!(ref self, PermissionControl);
+            assert(
+                permission_control_comp.has_proposer_permission(get_caller_address()),
+                Errors::ERR_NO_PROPOSER_PERMISSION
+            );
+            // Approve the account for future transfer
+            erc721_dispatcher.approve(account_address, token_id);
```
Comment (codebestia) [inline src/actions/nft_transaction.cairo:73]: No need to check for proposer permission. The create_transaction function does that already.
Also remove the `erc721_dispatcher.approve(account_address, token_id)` line. This give the recipient authority to transfer the NFT from the account even if the transaction has not been approved and we don't want that.

File: src/actions/nft_transaction.cairo [hunk 2]
```diff
@@ -1,10 +1,389 @@
...
+                    let id = self.nft_transaction_ids.at(index).read();
+                    let tx = self.nft_transactions.entry(id).read();
+                    array.append(tx);
+                };
+            array
+        }
+    }
 }
+// #[starknet::component]
+// pub mod NFTTransaction {
```
Comment (codebestia) [inline src/actions/nft_transaction.cairo:116]: Remove the commented code snippet as it is not needed.
Lets prevent having a messy codebase.

File: src/interfaces/ierc721.cairo [hunk 3]
```diff
@@ -1 +1,31 @@
...
+        self: @TContractState, owner: ContractAddress, operator: ContractAddress
+    ) -> bool;
+
+    // IERC721Metadata
+    fn name(self: @TContractState) -> ByteArray;
+    fn symbol(self: @TContractState) -> ByteArray;
+    fn token_uri(self: @TContractState, token_id: u256) -> ByteArray;
+
+    // NFT contract
+    fn mint(ref self: TContractState, recipient: ContractAddress, token_id: u256);
```
Comment (codebestia) [inline src/interfaces/ierc721.cairo:30]: This is not a valid interface function. remove it please.

**Review by codebestia — APPROVED** [review]
Good job so far @FrankiePower 
Please attend to the requested changes and add tests.

Please send a completed implementation before 4th of May.

File: src/actions/nft_transaction.cairo (hunk 1 above)
Comment (FrankiePower) [reply src/actions/nft_transaction.cairo:73]: noted

**Comment by codebestia** [comment]
Good job @FrankiePower. 
Everything looks good!
Please format your code and then this PR will be ready to be merged.

**Comment by FrankiePower** [comment]
@codebestia there is an issue. i am not yet done with the tests, because for some reason unknown to me it's not seeing it. you can take a look at my last push. i created a test file in the actions test folder, wrote some tests and tried to run it. but it doesnt work. i used this command: scarb test test_nft_transaction, i have also tried snforge test. what could be wrong ?

**Comment by codebestia** [comment]
> @codebestia there is an issue. i am not yet done with the tests, because for some reason unknown to me it's not seeing it. you can take a look at my last push. i created a test file in the actions test folder, wrote some tests and tried to run it. but it doesnt work. i used this command: scarb test test_nft_transaction, i have also tried snforge test. what could be wrong ?

Add the test file that you created to `lib.cairo`. Add the mock_nft file there as well.

**Comment by FrankiePower** [comment]
yes that did it. okay let me fix the failing tests.

**Comment by FrankiePower** [comment]
@codebestia i am done chief

**Comment by codebestia** [comment]
@FrankiePower 
Format the codebase.
The PR looks ready for merge

**Comment by FrankiePower** [comment]
@codebestia Done!!

**Comment by codebestia** [comment]
Please remove the `vscode` folder.

**Comment by FrankiePower** [comment]
@codebestia done!

**Review by codebestia — APPROVED** [review]
LGTM!
Thank you for your contribution.
